File: road/serde/json_serde.go
package serde

import "github.com/lixianmin/got/convert"

/********************************************************************
created:    2023-06-06
author:     lixianmin

Copyright (C) - All Rights Reserved
*********************************************************************/

type JsonSerde struct {
}

func (s *JsonSerde) Serialize(v any) ([]byte, error) {
	if bts, ok := v.([]byte); ok {
		return bts, nil
	}

	if v != nil {
		return convert.ToJsonE(v)
	}

	return nil, nil
}

func (s *JsonSerde) Deserialize(data []byte, v any) error {
	return convert.FromJsonE(data, v)
}

File: web/src/code/road/json_serde.js
'use strict'

/********************************************************************
 created:    2023-06-09
 author:     lixianmin

 Copyright (C) - All Rights Reserved
 *********************************************************************/

export function newJsonSerde() {
    const encoder = new TextEncoder()
    const decoder = new TextDecoder()

    function serialize(v) {
        const text = JSON.stringify(v)
        return encoder.encode(text)
    }

    function deserialize(bytes) {
        const text = decoder.decode(bytes)
        return JSON.parse(text)
    }

    function bytes2String(bytes) {
        return decoder.decode(bytes)
    }

    function string2bytes(s) {
        return encoder.encode(s)
    }

    return {
        serialize: serialize,
        deserialize: deserialize,
        bytes2String: bytes2String,
        string2bytes: string2bytes,
    }
}